In [108]:
import pandas as pd

In [109]:
data=pd.read_csv(r"C:\Users\hp\Downloads\adult Data Set.csv",skipinitialspace=True)

In [110]:
data

age         workclass  fnlwgt   education  education-num  \
0       39         State-gov   77516   Bachelors             13   
1       50  Self-emp-not-inc   83311   Bachelors             13   
2       38           Private  215646     HS-grad              9   
3       53           Private  234721        11th              7   
4       28           Private  338409   Bachelors             13   
...    ...               ...     ...         ...            ...   
32556   27           Private  257302  Assoc-acdm             12   
32557   40           Private  154374     HS-grad              9   
32558   58           Private  151910     HS-grad              9   
32559   22           Private  201490     HS-grad              9   
32560   52      Self-emp-inc  287927     HS-grad              9   

           marital-status         occupation   relationship   race     sex  \
0           Never-married       Adm-clerical  Not-in-family  White    Male   
1      Married-civ-spouse    Exec-managerial        Husband  White    Male   
2                Divorced  Handlers-cleaners  Not-in-family  White    Male   
3      Married-civ-spouse  Handlers-cleaners        Husband  Black    Male   
4      Married-civ-spouse     Prof-specialty           Wife  Black  Female   
...                   ...                ...            ...    ...     ...   
32556  Married-civ-spouse       Tech-support           Wife  White  Female   
32557  Married-civ-spouse  Machine-op-inspct        Husband  White    Male   
32558             Widowed       Adm-clerical      Unmarried  White  Female   
32559       Never-married       Adm-clerical      Own-child  White    Male   
32560  Married-civ-spouse    Exec-managerial           Wife  White  Female   

       capital-gain  capital-loss  hours-per-week native-country Salary_range  
0              2174             0              40  United-States        <=50K  
1                 0             0              13  United-States        <=50K  
2                 0             0              40  United-States        <=50K  
3                 0             0              40  United-States        <=50K  
4                 0             0              40           Cuba        <=50K  
...             ...           ...             ...            ...          ...  
32556             0             0              38  United-States        <=50K  
32557             0             0              40  United-States         >50K  
32558             0             0              40  United-States        <=50K  
32559             0             0              20  United-States        <=50K  
32560         15024             0              40  United-States         >50K  

[32561 rows x 15 columns]

In [111]:
data.rename(columns={'native-country':'native'},inplace=True)
data

age         workclass  fnlwgt   education  education-num  \
0       39         State-gov   77516   Bachelors             13   
1       50  Self-emp-not-inc   83311   Bachelors             13   
2       38           Private  215646     HS-grad              9   
3       53           Private  234721        11th              7   
4       28           Private  338409   Bachelors             13   
...    ...               ...     ...         ...            ...   
32556   27           Private  257302  Assoc-acdm             12   
32557   40           Private  154374     HS-grad              9   
32558   58           Private  151910     HS-grad              9   
32559   22           Private  201490     HS-grad              9   
32560   52      Self-emp-inc  287927     HS-grad              9   

           marital-status         occupation   relationship   race     sex  \
0           Never-married       Adm-clerical  Not-in-family  White    Male   
1      Married-civ-spouse    Exec-managerial        Husband  White    Male   
2                Divorced  Handlers-cleaners  Not-in-family  White    Male   
3      Married-civ-spouse  Handlers-cleaners        Husband  Black    Male   
4      Married-civ-spouse     Prof-specialty           Wife  Black  Female   
...                   ...                ...            ...    ...     ...   
32556  Married-civ-spouse       Tech-support           Wife  White  Female   
32557  Married-civ-spouse  Machine-op-inspct        Husband  White    Male   
32558             Widowed       Adm-clerical      Unmarried  White  Female   
32559       Never-married       Adm-clerical      Own-child  White    Male   
32560  Married-civ-spouse    Exec-managerial           Wife  White  Female   

       capital-gain  capital-loss  hours-per-week         native Salary_range  
0              2174             0              40  United-States        <=50K  
1                 0             0              13  United-States        <=50K  
2                 0             0              40  United-States        <=50K  
3                 0             0              40  United-States        <=50K  
4                 0             0              40           Cuba        <=50K  
...             ...           ...             ...            ...          ...  
32556             0             0              38  United-States        <=50K  
32557             0             0              40  United-States         >50K  
32558             0             0              40  United-States        <=50K  
32559             0             0              20  United-States        <=50K  
32560         15024             0              40  United-States         >50K  

[32561 rows x 15 columns]

In [45]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32561 entries, 0 to 32560
Data columns (total 15 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   age             32561 non-null  int64 
 1   workclass       32561 non-null  object
 2   fnlwgt          32561 non-null  int64 
 3   education       32561 non-null  object
 4   education-num   32561 non-null  int64 
 5   marital-status  32561 non-null  object
 6   occupation      32561 non-null  object
 7   relationship    32561 non-null  object
 8   race            32561 non-null  object
 9   sex             32561 non-null  object
 10  capital-gain    32561 non-null  int64 
 11  capital-loss    32561 non-null  int64 
 12  hours-per-week  32561 non-null  int64 
 13  native          32561 non-null  object
 14  Salary_range    32561 non-null  object
dtypes: int64(6), object(9)
memory usage: 3.7+ MB


In [46]:
data.isnull().sum()

age               0
workclass         0
fnlwgt            0
education         0
education-num     0
marital-status    0
occupation        0
relationship      0
race              0
sex               0
capital-gain      0
capital-loss      0
hours-per-week    0
native            0
Salary_range      0
dtype: int64

In [47]:
data['age'].isnull().sum()

0

In [48]:
#Q1.Shape of dataset 

In [49]:
data.shape

(32561, 15)

In [50]:
#Q2.Print column names 

In [51]:
data.columns

Index(['age', 'workclass', 'fnlwgt', 'education', 'education-num',
       'marital-status', 'occupation', 'relationship', 'race', 'sex',
       'capital-gain', 'capital-loss', 'hours-per-week', 'native',
       'Salary_range'],
      dtype='object')

In [52]:
#Q3.How many men and women (sex feature) are represented in this dataset? 

In [53]:
data.sex.value_counts()

Male      21790
Female    10771
Name: sex, dtype: int64

In [54]:
#Q4.What is the average age (age feature) of women? 

In [55]:
women=data[data['sex']=='Female']['age'].mean()
women

36.85823043357163

In [56]:
 #Q5.What is the percentage of German citizens (native-country feature)? 

In [112]:
x=(data['native']=='Germany').sum()
x

137

In [118]:
len(data)

32561

In [121]:
y=(x/len(data))*100
y

0.42074874850281013

The Percentage Of German citizens is 0.42%

In [60]:
#Q6.What are the mean and standard deviation of age for those who earn more than 50K per year (salary feature) and those who earn less than 50K per year? 

In [61]:
data[data['Salary_range']>'50K']

age         workclass  fnlwgt   education  education-num  \
0       39         State-gov   77516   Bachelors             13   
1       50  Self-emp-not-inc   83311   Bachelors             13   
2       38           Private  215646     HS-grad              9   
3       53           Private  234721        11th              7   
4       28           Private  338409   Bachelors             13   
...    ...               ...     ...         ...            ...   
32556   27           Private  257302  Assoc-acdm             12   
32557   40           Private  154374     HS-grad              9   
32558   58           Private  151910     HS-grad              9   
32559   22           Private  201490     HS-grad              9   
32560   52      Self-emp-inc  287927     HS-grad              9   

           marital-status         occupation   relationship   race     sex  \
0           Never-married       Adm-clerical  Not-in-family  White    Male   
1      Married-civ-spouse    Exec-managerial        Husband  White    Male   
2                Divorced  Handlers-cleaners  Not-in-family  White    Male   
3      Married-civ-spouse  Handlers-cleaners        Husband  Black    Male   
4      Married-civ-spouse     Prof-specialty           Wife  Black  Female   
...                   ...                ...            ...    ...     ...   
32556  Married-civ-spouse       Tech-support           Wife  White  Female   
32557  Married-civ-spouse  Machine-op-inspct        Husband  White    Male   
32558             Widowed       Adm-clerical      Unmarried  White  Female   
32559       Never-married       Adm-clerical      Own-child  White    Male   
32560  Married-civ-spouse    Exec-managerial           Wife  White  Female   

       capital-gain  capital-loss  hours-per-week         native Salary_range  
0              2174             0              40  United-States        <=50K  
1                 0             0              13  United-States        <=50K  
2                 0             0              40  United-States        <=50K  
3                 0             0              40  United-States        <=50K  
4                 0             0              40           Cuba        <=50K  
...             ...           ...             ...            ...          ...  
32556             0             0              38  United-States        <=50K  
32557             0             0              40  United-States         >50K  
32558             0             0              40  United-States        <=50K  
32559             0             0              20  United-States        <=50K  
32560         15024             0              40  United-States         >50K  

[32561 rows x 15 columns]

1. Mean Age of those , whose salary>50k

In [62]:
data[data['Salary_range']=='>50K']['age'].mean()

44.24984058155847

2. Mean Age of those , whose salary<=50k

In [63]:
data[data['Salary_range']=='<=50K']['age'].mean()

36.78373786407767

  3.Standard deviation of Age >50k 

In [64]:
data[data['Salary_range']=='>50K']['age'].std()

10.51902771985177

4.Standard deviation of Age<=50k

In [122]:
data[data['Salary_range']=='<=50K']['age'].std()

14.020088490824813

In [123]:
#Q7.Is it true that people who earn more than 50K have at least high school education? (education – Bachelors, Prof-school, Assoc-acdm, Assoc-voc, Masters or Doctorate feature) 

In [124]:
data.education.unique()

array(['Bachelors', 'HS-grad', '11th', 'Masters', '9th', 'Some-college',
       'Assoc-acdm', 'Assoc-voc', '7th-8th', 'Doctorate', 'Prof-school',
       '5th-6th', '10th', '1st-4th', 'Preschool', '12th'], dtype=object)

In [137]:
data[data['Salary_range']=='>50K']['education'].unique()

array(['HS-grad', 'Masters', 'Bachelors', 'Some-college', 'Assoc-voc',
       'Doctorate', 'Prof-school', 'Assoc-acdm', '7th-8th', '12th',
       '10th', '11th', '9th', '5th-6th', '1st-4th'], dtype=object)

In [134]:
data[(data['Salary_range']=='>50K')]['education'].value_counts()

Bachelors       2221
HS-grad         1675
Some-college    1387
Masters          959
Prof-school      423
Assoc-voc        361
Doctorate        306
Assoc-acdm       265
10th              62
11th              60
7th-8th           40
12th              33
9th               27
5th-6th           16
1st-4th            6
Name: education, dtype: int64

In [126]:
data[data['Salary_range']=='<=50K']['education'].unique()

array(['Bachelors', 'HS-grad', '11th', 'Masters', '9th', 'Assoc-acdm',
       '7th-8th', 'Some-college', 'Assoc-voc', '5th-6th', '10th',
       '1st-4th', 'Preschool', '12th', 'Doctorate', 'Prof-school'],
      dtype=object)

In [136]:
data[(data['Salary_range']=='<=50K')]['education'].value_counts()

HS-grad         8826
Some-college    5904
Bachelors       3134
11th            1115
Assoc-voc       1021
10th             871
Assoc-acdm       802
Masters          764
7th-8th          606
9th              487
12th             400
5th-6th          317
1st-4th          162
Prof-school      153
Doctorate        107
Preschool         51
Name: education, dtype: int64

NO Is it  NOT true that people who earn more than 50K have at least high school education

In [70]:
#Q8.Display age statistics for each race (race feature) and each gender (sex feature). Use groupby() and describe(). Find the maximum age of men of Amer-Indian-Eskimo race. 

In [71]:
b=data.groupby(['race','sex'])['age'].describe()
b

count       mean        std   min   25%   50%  \
race               sex                                                       
Amer-Indian-Eskimo Female    119.0  37.117647  13.114991  17.0  27.0  36.0   
                   Male      192.0  37.208333  12.049563  17.0  28.0  35.0   
Asian-Pac-Islander Female    346.0  35.089595  12.300845  17.0  25.0  33.0   
                   Male      693.0  39.073593  12.883944  18.0  29.0  37.0   
Black              Female   1555.0  37.854019  12.637197  17.0  28.0  37.0   
                   Male     1569.0  37.682600  12.882612  17.0  27.0  36.0   
Other              Female    109.0  31.678899  11.631599  17.0  23.0  29.0   
                   Male      162.0  34.654321  11.355531  17.0  26.0  32.0   
White              Female   8642.0  36.811618  14.329093  17.0  25.0  35.0   
                   Male    19174.0  39.652498  13.436029  17.0  29.0  38.0   

                             75%   max  
race               sex                  
Amer-Indian-Eskimo Female  46.00  80.0  
                   Male    45.00  82.0  
Asian-Pac-Islander Female  43.75  75.0  
                   Male    46.00  90.0  
Black              Female  46.00  90.0  
                   Male    46.00  90.0  
Other              Female  39.00  74.0  
                   Male    42.00  77.0  
White              Female  46.00  90.0  
                   Male    49.00  90.0

In [72]:
data.loc[(data['race']=='Amer-Indian-Eskimo') & (data['sex']=='Male')]['age'].max()

82

The maximum age of men of Amer-Indian-Eskimo race Is 82 Year

In [73]:
b.iloc[1:2]

count       mean        std   min   25%   50%   75%  \
race               sex                                                         
Amer-Indian-Eskimo Male  192.0  37.208333  12.049563  17.0  28.0  35.0  45.0   

                          max  
race               sex         
Amer-Indian-Eskimo Male  82.0

In [74]:
#Q9.Among whom is the proportion of those who earn a lot (>50K) greater: married or single men (marital-status feature)? Consider as married those who have a marital-status starting with  Married (Married-civ-spouse, Married-spouse-absent or Married-AF-spouse), the rest are considered bachelors. 

In [75]:
data['marital-status'].unique()

array(['Never-married', 'Married-civ-spouse', 'Divorced',
       'Married-spouse-absent', 'Separated', 'Married-AF-spouse',
       'Widowed'], dtype=object)

In [76]:
married=data.loc[(data['sex']=='Male') & (data['marital-status'].isin(['Married-spouse-absent','Married-civ-spouse','Married-AF-spouse'])),'Salary_range'].value_counts()
married

<=50K    7576
>50K     5965
Name: Salary_range, dtype: int64

In [77]:
single=data.loc[(data['sex']=='Male') & (data['marital-status'].isin(['Never-married','Divorced','Separated','Widowed'])),'Salary_range'].value_counts()
single

<=50K    7552
>50K      697
Name: Salary_range, dtype: int64

From above count it is clearly seen that there is large number of 
married people whose salary is greater than 50K than single peoples.

In [78]:
#Q10.What is the maximum number of hours a person works per week

maximum number of hours a person works per week

In [79]:
data['hours-per-week'].max()

99

Number of people who work max 99 hour per week

In [143]:
count=(data['hours-per-week']==99).sum()
count

85

Number of People who work max 99 hour per week and earn greater than 50K

In [150]:
Count=data.loc[(data['hours-per-week']==99) & (data['Salary_range']=='>50K')]['Salary_range'].value_counts()/count*100

Count

>50K    29.411765
Name: Salary_range, dtype: float64

Percentage of people who earn greater than 50k is 29.41%

In [85]:
#Q11.Count the average time of work (hours-per-week) for those who earn a little and a lot (salary) for each country (native-country). What will these be for Japan? 

In [86]:
data1=data.loc[data['Salary_range']=='>50K']

In [94]:
data1.groupby(['native'])[['hours-per-week']].describe()


hours-per-week                                           \
                            count       mean        std   min    25%   50%   
native                                                                       
?                           146.0  45.547945  10.202827   8.0  40.00  40.0   
Cambodia                      7.0  40.000000   0.000000  40.0  40.00  40.0   
Canada                       39.0  45.641026  12.066673  20.0  40.00  45.0   
China                        20.0  38.900000  10.125008  10.0  38.00  40.0   
Columbia                      2.0  50.000000   0.000000  50.0  50.00  50.0   
Cuba                         25.0  42.440000  12.453514  14.0  40.00  40.0   
Dominican-Republic            2.0  47.000000   9.899495  40.0  43.50  47.0   
Ecuador                       4.0  48.750000   2.500000  45.0  48.75  50.0   
El-Salvador                   9.0  45.000000   7.071068  35.0  40.00  45.0   
England                      30.0  44.533333  11.072592  20.0  40.00  40.0   
France                       12.0  50.750000  16.360290  40.0  40.00  47.5   
Germany                      44.0  44.977273  11.944618   8.0  40.00  45.0   
Greece                        8.0  50.625000  11.783008  35.0  40.00  50.0   
Guatemala                     3.0  36.666667   5.773503  30.0  35.00  40.0   
Haiti                         4.0  42.750000   9.142392  35.0  38.75  40.0   
Honduras                      1.0  60.000000        NaN  60.0  60.00  60.0   
Hong                          6.0  45.000000  10.000000  40.0  40.00  40.0   
Hungary                       3.0  50.000000  10.000000  40.0  45.00  50.0   
India                        40.0  46.475000  10.056220  20.0  40.00  41.5   
Iran                         18.0  47.500000  10.325582  30.0  40.00  45.0   
Ireland                       5.0  48.000000  10.954451  40.0  40.00  40.0   
Italy                        25.0  45.400000   9.495613  20.0  40.00  40.0   
Jamaica                      10.0  41.100000   4.954235  35.0  38.50  40.0   
Japan                        24.0  47.958333  16.120414  21.0  40.00  42.5   
Laos                          2.0  40.000000   0.000000  40.0  40.00  40.0   
Mexico                       33.0  46.575758  10.917390  30.0  40.00  40.0   
Nicaragua                     2.0  37.500000   3.535534  35.0  36.25  37.5   
Peru                          2.0  40.000000   0.000000  40.0  40.00  40.0   
Philippines                  61.0  43.032787   8.445842  35.0  40.00  40.0   
Poland                       12.0  39.000000   9.779385  10.0  40.00  40.0   
Portugal                      4.0  41.500000   5.972158  36.0  39.00  40.0   
Puerto-Rico                  12.0  39.416667  11.057357   8.0  40.00  40.0   
Scotland                      3.0  46.666667  11.547005  40.0  40.00  40.0   
South                        16.0  51.437500  10.942082  35.0  40.00  50.0   
Taiwan                       20.0  46.800000  13.485275  27.0  40.00  40.0   
Thailand                      3.0  58.333333  14.433757  50.0  50.00  50.0   
Trinadad&Tobago               2.0  40.000000   0.000000  40.0  40.00  40.0   
United-States              7171.0  45.505369  11.025092   1.0  40.00  40.0   
Vietnam                       5.0  39.200000   1.788854  36.0  40.00  40.0   
Yugoslavia                    6.0  49.500000  11.202678  40.0  40.00  46.0   

                                 
                      75%   max  
native                           
?                   50.00  80.0  
Cambodia            40.00  40.0  
Canada              51.00  72.0  
China               41.25  60.0  
Columbia            50.00  50.0  
Cuba                50.00  60.0  
Dominican-Republic  50.50  54.0  
Ecuador             50.00  50.0  
El-Salvador         50.00  55.0  
England             50.00  85.0  
France              50.00  99.0  
Germany             50.00  70.0  
Greece              61.25  65.0  
Guatemala           40.00  40.0  
Haiti               44.00  56.0  
Honduras            60.00  60.0  
Hong              

In [100]:
data.loc[(data['native']=='Japan') &  (data['Salary_range']=='>50K')]['hours-per-week'].mean()

47.958333333333336

In [336]:
data2.loc[' Japan']

hours-per-week  count    24.000000
                mean     47.958333
                std      16.120414
                min      21.000000
                25%      40.000000
                50%      42.500000
                75%      50.000000
                max      99.000000
Name:  Japan, dtype: float64